fire/io/bernese/bernese.py

```diff
@@ -0,0 +1,351 @@
+    linjer af fast bredde
+
+    Linjerne følger efter denne header:
+    Sol Station name         Typ Correction  Estimated value  RMS error   A priori value Unit    From                To                  MJD           Num Abb
+
+    Eks.:
+
+      1 BOR1                   X    0.00216    3738358.26267    0.00035    3738358.26051 meters  2016-03-03 00:00:00 2016-03-04 23:59:30 57450.99983     1 #CRD
+      1 BOR1                   Y    0.00056    1148173.88492    0.00015    1148173.88436 meters  2016-03-03 00:00:00 2016-03-04 23:59:30 57450.99983     2 #CRD
+      1 BOR1                   Z    0.00187    5021815.87371    0.00046    5021815.87184 meters  2016-03-03 00:00:00 2016-03-04 23:59:30 57450.99983     3 #CRD
+      1 BUDP                   X    0.00011    3513638.07857    0.00034    3513638.07846 meters  2016-03-03 00:00:00 2016-03-04 23:59:30 57450.99983     4 #CRD
+      1 BUDP                   Y    0.00020     778956.56481    0.00015     778956.56461 meters  2016-03-03 00:00:00 2016-03-04 23:59:30 57450.99983     5 #CRD
+
+    """
+    station = line[5:9]
+    obsfra = line[94:113]
+    obstil = line[114:133]
+    return dict(zip_longest(["STATION NAME", "FROM", "TO"], [station, obsfra, obstil]))
+
+
+def addneq_parse_spreadline(line: str) -> dict:
+    """
+    Parse en linje med RMS-spredning fra en ADDNEQ-fil, udtræk stationsnavn, samt retning (N/E/U) og spredning.
+    En serie linjer kan se således ud:
+
+    GESR             N    0.07      0.02   -0.06
+    GESR             E    0.10     -0.00   -0.10
+    GESR             U    0.23     -0.10    0.20
+
+     - og det er de tre første 'ord' vi skal bruge - spredning er det første tal, de to andre ignorerer vi
+
+    """
+    params = line.split()
+    return dict(zip_longest(["STATION NAME", "DIRECTION", "SPREAD"], params))
+
+
+def cov_parse_dataline(line: str) -> dict:
+    """
+    Parse datasektionen fra en COV-linje og udtræk observationsfelterne - station 1 og 2, deres koordinatpar-labels og matriceelement.
+
+    En sektion kan begynde sådan her:
+
+    STATION 1        XYZ    STATION 2        XYZ FLG    MATRIX ELEMENT
+
+    BOR1              X     BOR1              X         0.9326975128D-01
+
+    BOR1              Y     BOR1              X         0.2008472992D-01
+    BOR1              Y     BOR1              Y         0.1634563313D-01
+
+    BOR1              Z     BOR1              X         0.9670543502D-01
+    BOR1              Z     BOR1              Y         0.2236056402D-01
+    BOR1              Z     BOR1              Z         0.1607287392D+00
+
+    BUDP              X     BOR1              X        -0.1649457919D-02
+
+    """
+    params = line.split()
+    return dict(
+        zip_longest(
+            ["STATION 1", "XYZ 1", "STATION 2", "XYZ 2", "MATRIX ELEMENT"], params
+        )
+    )
+
+
+class BerneseSolution(dict):
+    resultatset: BerneseSet
```

Comment:
Bruges dettee klasse-medlem? Ellers skal den bare slettes;)

(Alle instanser får deres eget instans-medlem i `__init__(...)`, så alle senere referencer til `self.resultatset` er til instans-medlemmet.)

Reply:
Som jeg bemærker i denne [kommentar](https://github.com/Kortforsyningen/FIRE/pull/540#discussion_r785828174) er `resultatset` egentligt mest bare fyld. Medmindre der er et eller andet jeg har overset i hvert fald. Jeg ser gerne at interfacet her forsimples.

Reply:
Nå, GitHub er ikke for god til at linke direkte til den kommentar. Herunder en gentagelse:

> Jeg bryder mig ikke om den her konstruktion, da den egentligt bare er fyld. En `BerneseSolution` indeholder altid _et_ `BerneseSet`, som da holder styr på nogle få metadata og derudover har en dict med alle stationerne. Det resulterer i kald i stil med
> 
> ```
> b = BerneseSolution(...)
> b.resultatset.stations['BUDP'].koordinat.x
> b.resultatset.gnss_week
> ```
> 
> hvilket lige så godt kunne være
> 
> ```
> b = BerneseSolution(...)
> b['BUDP'].koordinat.x
> b.gnss_week
> ```
> 
> Det er både lettere at læse og skrive. Hvis du nedarver fra `dict` er det ret smertefrit at komme frem til den løsning. Her er et hurtigt eksempel der demonstrerer hvordan det fx kan gøres:
> 
> ```python
> class BerneseSolution(dict):
> 
>     def __init__(self, stationer, koordinater, referenceramme):
> 
>         for station, koordinat in zip(stationer, koordinater):
>             # Her fyldes
>             self[station] = koordinat
> 
>         self.referenceramme = referenceramme
> 
> 
> s = BerneseSolution(
>     ['BUDP', 'SMID', 'MOJN'],
>     [
>         (12.5000197700, 55.7390166857, 94.01580),
>         (9.5593019761, 55.6406426308, 122.81450),
>         (8.8053784657, 54.9443151050, 57.12874)
>     ],
>     "igb14",
> )
> 
> for station in s.items():
>     print(station)
> 
> print(s['BUDP'])
> print(s.referenceramme)
> ```

Reply:
Det er forsimplet nu. Oprindeligt lavede jeg for meget boilerplate kode, så nu skulle der være mindre af den slags.

Reply:
> > Det er både lettere at læse og skrive. Hvis du nedarver fra `dict` er det ret smertefrit at komme frem til den løsning. Her er et hurtigt eksempel der demonstrerer hvordan det fx kan gøres:
> > ```python
> > class BerneseSolution(dict):
> > 
> >     def __init__(self, stationer, koordinater, referenceramme):
> > 
> >         for station, koordinat in zip(stationer, koordinater):
> >             # Her fyldes
> >             self[station] = koordinat
> > 
> >         self.referenceramme = referenceramme
> > 
> > 
> > s = BerneseSolution(
> >     ['BUDP', 'SMID', 'MOJN'],
> >     [
> >         (12.5000197700, 55.7390166857, 94.01580),
> >         (9.5593019761, 55.6406426308, 122.81450),
> >         (8.8053784657, 54.9443151050, 57.12874)
> >     ],
> >     "igb14",
> > )
> > 
> > for station in s.items():
> >     print(station)
> > 
> > print(s['BUDP'])
> > print(s.referenceramme)
> > ```

Lige nu initieres `BerneseSolution` med parametrene for filstier. Skulle jeg istedet lave en `load(...)` funktion til den funktionalitet så?

Reply:
Ovenstående var barre et eksempel for at illustrere en pointe. Du behøver ikke lave en dedikeret load funktion, du er der jo stort set allerede. I `BerneseSolution.__init__()` starter du med

```
super().__init__()
self.stationer = {}
```

rundt omkring i de forskellige metoder har du så for eksempel

```
self.stationer[datalinje["STATION NAME"]] = station
```

Da du nedarver fra `dict` kan du simpelthen bare fjerne `.stationer`:
```
self[datalinje["STATION NAME"]] = station
```

og så kan du efterfølgende tilgå en station med `self[station]` i stedet for `self.stationer[station]`. Det er primært kosmetisk men det forsimpler da i hvert fald tingene for brugeren af en `BerneseSolution`.

Reply:
Det skulle jeg have implementeret nu :)